proposals/0000-compiler-settings.md

```diff
@@ -0,0 +1,279 @@
+   module basis allowing for warningsAsError to be used per file instead of per
+   module migration to use new features.
+
+As an additional benefit since enums are being used, code completion can make
+the compiler options discoverable to the user in the IDE.
+
+## Detailed design
+
+We will update the swift grammer to allow for a new top level statement called
+`compilerSettings` that takes a list of expressions:
+
+```text
+top-level-statement ::= 'compilerSettings' '[' (expr ',')* expr ','? ']'
+```
+
+During typechecking, we will validate that each of the expressions are instances
+of the enum `CompilerSetting`. `CompilerSetting` will be a resilient enum
+defined in the standard library that provides the following API:
+
+```
+public enum CompilerSetting {
+case enableUpcomingFeature(UpcomingFeatureCompilerSetting)
+case enableExperimentalFeature(ExperimentalFeatureCompilerSetting)
+case warningAsError(WarningAsErrorCompilerSetting)
+case defaultIsolation(Actor.Type)
+case strictConcurrency(StrictConcurrencyCompilerSetting)
+}
+```
+
+We purposely make `CompilerSetting` resilient so additional kinds of compiler
+settings can be added over time without breaking ABI.
+
+`CompilerSetting` provides an enum case for each command line option and each
+case is able to provide associated values to specify arguments that would
+normally be passed on the command line to the option. These associated values
+can be one of:
+
+* an already existing type (e.x.: `Actor.Type`)
+
+* a custom fixed enum defined as a subtype of `CompilerSetting` (e.x.:
+  `StrictConcurrencyCompilerSetting`)
+
+* a custom enum defined as a subtype of `CompilerSetting` that the compiler
+  synthesizes cases for depending on internal compiler data (e.x.: features for
+  `UpcomingFeatureCompilerSetting`)
+
+We purposely keep all custom types defined for `CompilerSetting` as subtypes of
+`CompilerSetting` in order to avoid polluting the global namespace. Thus we
+would necessarily define in the stdlib all of the custom types as:
+
+```swift
+extension CompilerSetting {
+  // Cases synthesized by the compiler.
+  public enum UpcomingFeatureCompilerSetting { }
+
+  // Cases synthesized by the compiler.
+  public enum ExperimentalFeatureCompilerSetting { }
+
+  // Cases synthesized by the compiler.
+  public enum WarningsAsErrorsCompilerSetting { }
+
+  // We know ahead of time all of the cases, so this is specified explicitly.
+  public enum StrictConcurrency {
+  case minimal
+  case targeted
+  case complete
+  }
+}
+```
+
+In order to ensure that compiler settings are easy to find in the file, we
+require that `compilerSettings` be the first AST node in the file before any
+other statements including import statements.
+
+After parsing, the compiler in the type checker validates that each element is a
+`CompilerSetting` and then updates the relevant compiler state as appropriate
+for the command line option.
```

Comment:
You don't need to mention parsing, type checking, compiler state, etc, here - those are compiler implementation details. I think all we need to say is that the argument to `compilerSetting` has type `[CompilerSetting]` and those settings are applied only to the file they're specified in.